src/data/real/pi.lean

```diff
@@ -186,42 +160,94 @@ begin
-  apply sqrt_two_add_series_step_up 1447 727,
-  all_goals {norm_num}
+  refine lt_of_lt_of_le (pi_lt_sqrt_two_add_series n) _,
+  rw [← le_sub_iff_add_le, ← le_div_iff', sqrt_le_left, sub_le],
+  { rwa [nat.cast_zero, zero_div] at h },
+  { exact div_nonneg (sub_nonneg.2 h₂) (pow_pos two_pos _) },
+  { exact pow_pos two_pos _ }
 end
 
-lemma pi_lt_315 : pi < 3.15 :=
+lemma sqrt_two_add_series_step_down (a b : ℕ) {c d n : ℕ} {z : ℝ}
+  (hz : z ≤ sqrt_two_add_series (a/b) n) (hb : 0 < b) (hd : 0 < d)
+  (h : a ^ 2 * d ≤ (2 * d + c) * b ^ 2) : z ≤ sqrt_two_add_series (c/d) (n+1) :=
 begin
-  refine lt_of_lt_of_le (pi_lt_sqrt_two_add_series 4) _,
-  apply add_le_of_le_sub_right, rw [mul_comm], apply mul_le_of_le_div, apply pow_pos, norm_num,
-  rw [sqrt_le_left, sub_le, show (0:ℝ) = (0:ℕ)/(1:ℕ), by rw [nat.cast_zero, zero_div]],
-  apply sqrt_two_add_series_step_down 140 99,
-  apply sqrt_two_add_series_step_down 279 151,
-  apply sqrt_two_add_series_step_down 51 26,
-  apply sqrt_two_add_series_step_down 412 207,
-  all_goals {norm_num}
+  apply le_trans hz, rw sqrt_two_add_series_succ, apply sqrt_two_add_series_monotone_left,
+  apply le_sqrt_of_sqr_le,
+  have hb' : 0 < (b:ℝ) := nat.cast_pos.2 hb,
+  have hd' : 0 < (d:ℝ) := nat.cast_pos.2 hd,
+  rw [div_pow, add_div_eq_mul_add_div _ _ (ne_of_gt hd'), div_le_div_iff (pow_pos hb' _) hd'],
+  exact_mod_cast h
 end
 
-/- A computation of the first 7 digits of pi is given here:
-  <https://gist.github.com/fpvandoorn/5b405988bc2e61953d56e3597db16ecf>
-  This is not included in mathlib, because of slow compilation time.
-  -/
+/-- Create a proof of `pi < a` for a fixed rational number `a`, given a witness, which is a
+sequence of rational numbers `sqrt 2 < r 1 < r 2 < ... < r n < 2` satisfying the property that
+`sqrt (2 + r i) ≥ r(i+1)`, where `r 0 = 0` and `sqrt (2 - r n) ≥ (a - 1/4^n) / 2^(n+1)`. -/
+meta def pi_upper_bound (l : list ℚ) : tactic unit :=
+do let n := l.length,
+  (() <$ tactic.apply `(@pi_upper_bound_start %%(reflect n))); [pure (), `[norm_num1]],
+  l.mmap' (λ r, do
+    let a := r.num.to_nat, let b := r.denom,
+    (() <$ tactic.apply `(@sqrt_two_add_series_step_down %%(reflect a) %%(reflect b)));
+    [pure (), `[norm_num1], `[norm_num1], `[norm_num1]]),
+  `[simp only [sqrt_two_add_series, nat.cast_bit0, nat.cast_bit1, nat.cast_one, nat.cast_zero]],
+  `[norm_num]
+
+lemma pi_gt_three : 3 < pi := by pi_lower_bound [23/16]
+
+lemma pi_gt_314 : 3.14 < pi := by pi_lower_bound [99/70, 874/473, 1940/989, 1447/727]
+
+lemma pi_lt_315 : pi < 3.15 := by pi_upper_bound [140/99, 279/151, 51/26, 412/207]
+
+lemma pi_gt_31415 : 3.1415 < pi := by pi_lower_bound [
+  11482/8119, 5401/2923, 2348/1197, 11367/5711, 25705/12868, 23235/11621]
+
+lemma pi_lt_31416 : pi < 3.1416 := by pi_upper_bound [
+  4756/3363, 101211/54775, 505534/257719, 83289/41846,
+  411278/205887, 438142/219137, 451504/225769, 265603/132804, 849938/424971]
+
+lemma pi_gt_3141592 : 3.141592 < pi := by pi_lower_bound [
```

Comment:
Doe we save time by first computing these last two bounds using `pi_lower_bound` and `pi_upper_bound` and then deducing the courser bounds using `lt_trans` and `norm_num`?

Reply:
Most probably. The other bounds arent really necessary except to show off some certificates.